File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        count = 0
        prefix_sum_count = defaultdict(int)
        prefix_sum_count[0] = 1
        prefix_sum = 0
        
        for num in nums:
            prefix_sum += num
            count += prefix_sum_count[prefix_sum - goal]
            prefix_sum_count[prefix_sum] += 1
        
        return count

File: 0930-binary-subarrays-with-sum/0930-binary-subarrays-with-sum.cpp
class Solution {
public:
        int numSubarraysWithSum(vector<int>& nums, int goal) {
        int count = 0;
        unordered_map<int, int> prefixSumCount;
        prefixSumCount[0] = 1;
        int prefixSum = 0;
        
        for (int num : nums) {
            prefixSum += num;
            count += prefixSumCount[prefixSum - goal];
            prefixSumCount[prefixSum]++;
        }
        
        return count;
    }
};